### PR TITLE
Fix 156

### LINKED ISSUE #156: Error at `component push -c fruit-client-sb`
Local changes detected for 'fruit-client-sb' component: about to push source code to remote cluster
```
✗  Uploading /home/iocanel/rhte/fruit-client-sb.tar
FATA[0001] error running push: error uploading file: exit status 1
```

### PR DESCRIPTION
**Pull request by metacosm**
- Improve status computation so that it is more precise and can take different cases into account. Status reporting should also be improved.
- In particular, we are now able to change the status of the associated component when a link is created, using a newly introduced `ComponentLinking` status. 
- This allows the controller to distinguish between a `regular pending state` and a `state` where the component `needs to wait` for its associated pod to be re-created: the component controller thus won't allow the component to switch back to ready until the associated pod has been recreated when the component is in the Linking state. The pod name will also be properly updated.
- This should properly fix #156 because it will cause `hal` to wait for the component to switch back to ready before attempting to push.

### CODE REVIEW

**Comment by cmoulliard** [comment]
When we execute this scenario using latest version of hal = 0.1.10 - https://gist.github.com/cmoulliard/9ea0351fd7848835f1b50335b3485e1f and that we check the link, we go bad status message
```
oc get links
NAME            AGE       STATUS    MESSAGE
backend-to-db   5m        Pending   Waiting for the following resources:
Component: Ready
client-to-backend   5m        Pending   Waiting for the following resources:
Component: Ready
```

File: go.mod [hunk 1]
```diff
@@ -26,12 +26,12 @@ require (
...
 	golang.org/x/net v0.0.0-20190724013045-ca1201d0de80
-	golang.org/x/tools v0.0.0-20190506145303-2d16b83fe98c // indirect
-	halkyon.io/api v1.0.0-beta.5
+	golang.org/x/tools v0.0.0-20190506145303-2d16b83fe98c
+	halkyon.io/api v1.0.0-beta.7
 	k8s.io/api v0.0.0-20190725062911-6607c48751ae
 	k8s.io/apimachinery v0.0.0-20190719140911-bfcf53abc9f8
 	k8s.io/client-go v11.0.0+incompatible
-	k8s.io/kubernetes v1.14.5 // indirect
+	k8s.io/kubernetes v1.14.5
```
Comment (cmoulliard) [inline go.mod]: - Is is the same version as used within the API ? 
- Should we align this k8s version with what is defined within the doc ? 
- Are we doing our tests against k8s 1.14 ?

File: pkg/controller/component.go [hunk 2]
```diff
@@ -79,10 +79,15 @@ func (in *Component) SetErrorStatus(err error) bool {
...
 
-func (in *Component) SetSuccessStatus(dependentName, msg string) bool {
-	if dependentName != in.Status.PodName || halkyon.ComponentReady != in.Status.Phase || msg != in.Status.Message {
+func (in *Component) DependentStatusFieldName() string {
+	_ = in.Status.PodName // to make sure we update the value below if that field changes as returned value must match field name
+	return "PodName"
+}
+
+func (in *Component) SetSuccessStatus(statuses []DependentResourceStatus, msg string) bool {
+	// todo: compute message based on linking statuses
```
Comment (cmoulliard) [inline pkg/controller/component.go]: Is the TODO still needed ?

File: pkg/controller/component/controller.go [hunk 3]
```diff
@@ -158,22 +143,47 @@ func (r *ReconcileComponent) Delete(resource controller2.Resource) error {
...
-	}
-	return false
-}
-
-// Check if the TaskRun Condition is Type = SUCCEEDED and Status = True
-func (r *ReconcileComponent) isBuildSucceed(taskRun *taskRunv1alpha1.TaskRun) bool {
-	for _, c := range taskRun.Status.Conditions {
-		if c.Type == apis.ConditionSucceeded && c.Status == corev1.ConditionTrue {
-			return true
+func (r *ReconcileComponent) SetPrimaryResourceStatus(primary controller2.Resource, statuses []controller2.DependentResourceStatus) (needsUpdate bool) {
```
Comment (cmoulliard) [inline pkg/controller/component/controller.go]: Is it documented what are the component statuses and status about the dependent resources ?

File: pkg/controller/component.go (hunk 2 above)
Comment (metacosm) [reply pkg/controller/component.go]: Yes, it was just added. :)

File: pkg/controller/component/controller.go (hunk 3 above)
Comment (metacosm) [reply pkg/controller/component/controller.go]: Nothing is documented at this point.

File: go.mod (hunk 1 above)
Comment (metacosm) [reply go.mod]: I'm actually not sure where this is coming from… :(

**Pull request merged**